L74-java-channel-creds.md

```diff
@@ -0,0 +1,478 @@
+returned to produce a useful error message.
+
+`TlsChannelCredentials` will support a no-arg constructor for the "all
+defaults" configuration. For any more advanced configuration, a builder would
+be used.
+
+```java
+package io.grpc;
+
+public final class TlsChannelCredentials extends ChannelCredentials {
+  /**
+   * Returns an empty set if this credential can be adequately understood via
+   * the features listed, otherwise returns a hint of features that are lacking
+   * to understand the configuration to be used for manual debugging.
+   *
+   * <p>An "understood" feature does not imply the caller is able to fully
+   * handle the feature. It simply means the caller understands the feature
+   * enough to use the appropriate APIs to read the configuration. The caller
+   * may support just a subset of a feature, in which case the caller would
+   * need to look at the configuration to determine if only the supported
+   * subset is used.
+   *
+   * <p>This method may not be as simple as a set difference. There may be
+   * multiple features that can independently satisfy a piece of configuration.
+   * If the configuration is incomprehensible, all such features would be
+   * returned, even though only one may be necessary.
+   *
+   * <p>An empty set does not imply that the credentials are fully understood.
+   * There may be optional configuration that can be ignored if not understood.
+   */
+  public Set<Feature> incomprehensible(EnumSet<Feature> understoodFeatures) {...}
+
+  /* Other methods can be added to support various Features */
+
+  public enum Feature {}
+
+  public final static class Builder {
+    public TlsChannelCredentials build() {...}
+  }
+}
+```
+
+To support very advanced use cases, Netty will provide a credential that wraps
+a `ProtocolNegotiator`. This allows implementations like ALTS and XDS to use
+internal APIs without forcing their users to use experimental or internal APIs,
+as their users would just interact with `ChannelCredentials`.
+
+```java
+package io.grpc.netty;
+
+final class NettyChannelCredentials extends ChannelCredentials {
+  public NettyChannelCredentials(ProtocolNegotiator.Factory negotiator) {...}
+  public ProtocolNegotiator.Factory getNegotiator() { return negotiator; }
+}
+
+@Internal
+public final class InternalNettyChannelCredentials {
+  private InternalNettyChannelCredentials() {}
+
+  public ChannelCredentials create(InternalProtocolNegotiator.Factory negotiator)
+    {...}
+
+  /**
+   * Converts credentials to a negotiator, in similar fashion as for a new
+   * channel.
+   *
+   * @throws IllegalArgumentException if unable to convert
+   */
+  public InternalProtocolNegotiator.Factory toNegotiator(ChannelCredentials creds)
+    {...}
+}
+```
+
+### GoogleDefault ChannelCredentials Example
+
+The API provided here is not part of the design. It is meant as a
+demonstration.
+
+```java
+public final class GoogleDefaultChannelCredentials {
```

Comment:
thought: it's a little surprising to have the factory class called Credentials but not be a Credentials.

(I realize of course this is just an example).